be/src/runtime/data_stream_recvr.cc

```diff
@@ -441,19 +460,35 @@ Status DataStreamRecvr::SenderQueue::add_chunks_and_keep_order(const PTransmitCh
     size_t total_chunk_bytes = 0;
     faststring uncompressed_buffer;
     ChunkQueue local_chunk_queue;
-    for (auto& pchunk : request.chunks()) {
-        int64_t chunk_bytes = pchunk.data().size();
-        ChunkUniquePtr chunk = std::make_unique<vectorized::Chunk>();
-        RETURN_IF_ERROR(_deserialize_chunk(pchunk, chunk.get(), &uncompressed_buffer));
 
-        ChunkItem item{chunk_bytes, std::move(chunk), nullptr};
+    if (use_pass_through) {
+        ChunkUniquePtrVector swap_chunks;
+        std::vector<size_t> swap_bytes;
+        _recvr->_pass_through_context.pull_chunks(request.sender_id(), &swap_chunks, &swap_bytes);
+        DCHECK(swap_chunks.size() == swap_bytes.size());
+        size_t bytes = 0;
+        for (size_t i = 0; i < swap_chunks.size(); i++) {
+            ChunkItem item{static_cast<int64_t>(swap_bytes[i]), std::move(swap_chunks[i]), nullptr};
+            local_chunk_queue.emplace_back(std::move(item));
+            bytes += swap_bytes[i];
+        }
+        total_chunk_bytes += bytes;
+        COUNTER_UPDATE(_recvr->_bytes_pass_through_counter, total_chunk_bytes);
```

Comment:
Since chunks are pulled from _pass_through_context, why still use bprc?

Reply:
The design of this PR is to separate control flow from data flow. Control flow is still by brpc, and we optimzie data flow if sender/receiver in the same process. And we want to save the cost of chunk serialization and deserialization.

If we go further to eliminate brpc, we won't  save too much bandwidth or latency, and code will be complicated. So it does not worth of it.